webdriver-spec.html

```diff
@@ -1435,288 +1435,513 @@
 
- <li>If <var>desired capabilities</var> is not a JSON Object,
+  <ol>
+  <li><p>If <var>required capabilities</var> is <a>null</a> or <a>undefined</a>,
   set the value to an empty JSON Object.
 
- <li><p>Let <var>length</var> be the length of <var>required capabilities</var>.
-
- <li><p>Let <var>k</var> be 0.
+  <li><p>If <var>required capabilities</var> is not a JSON Object,
+  return <a>error</a> with <a>error code</a> <a>invalid argument</a>.
+  </ol>
 
- <li><p>While <var>k</var> &lt; <var>length</var>:
+<li><p>Let <var>desired capabilities</var> be the result of getting a property
+named "<code>firstMatch</code>" from <var>capabilities request</var>.
 
   <ol>
-   <li><p>Let <var>capability</var> be the entry
-    in <var>required capabilities</var> at index <var>k</var>.
-
-   <li><p>If the name of the <var>capability</var> is among
-    the names of entries in <var>desired capabilities</var>,
-    remove the corresponding entry from <var>desired capabilities</var>.
+  <li><p>If <var>desired capabilities</var> is <a>null</a> or <a>undefined</a>,
+  set the value to an empty JSON List.
 
-   <li>Increase <var>k</var> by 1.
+  <li><p>If <var>desired capabilities</var> is not a JSON List,
+  return <a>error</a> with <a>error code</a> <a>invalid argument</a>.
   </ol>
 
- <li><p>Let <var>unmet capabilities</var> be an empty JSON List.
-
- <li><p>Let <var>unprocessed capabilities</var> be a JSON Object
-  that contains all entries from <var>required capabilities</var>
-  and all entries from <var>desired capabilities</var>.
+<li><p>Let <var>length</var> be the length of <var>desired capabilities</var>.
 
- <li><p>Let <var>j</var> be 0.
+<li><p>Let <var>k</var> be 0.
 
- <li><p>Let <var>capabilities length</var> be
-  the length of <var>unprocessed capabilities</var>.
+<li><p>Let <var>matched capabilities</var> be <a>null</a>.
 
- <li><p>While <var>j</var> &lt; <var>capabilities length</var>:
+<li><p>While <var>k</var> &lt; <var>length</var> and
+<var>matched capabilities</var> is <a>null</a>:
 
   <ol>
-   <li><p>Let <var>unprocessed capability</var> be the entry at index <var>j</var>
-    in <var>unprocessed capabilities</var>.
+  <li><p>Let <var>dc</var> be the entry in <var>desired capabilities</var> at
+  index <var>k</var>.
 
-   <li><p>If during the steps below the <var>unprocessed capability</var>
-    equals an entry in <var>required capabilities</var>
-    and name of <var>unprocessed capability</var> entry among the names
-    of entries in <var>server capabilities</var>
-    and the values do not match do the following:
+  <li><p>If <var>dc</var> is not a JSON Object,
+  return <a>error</a> with <a>error code</a> <a>invalid argument</a>.
 
-    <ol>
-     <li><p>Append a string containing the property name
-      and the differences between the values.
-    </ol>
+  <li><p>Let <var>merged capabilities</var> be the result of
+  <a>merging capabilities</a> with <var>required capabilities</var> and
```

Comment:
Which takes precedence: the capabilities from "required" or "desired" capabilities. I'd suggest those from "required", but returning an error when keys clash would also be valid.